### PR TITLE
Change the order in which events are issued

### PR DESCRIPTION
**Pull request by heowc**
The order in which events are issued should be after the desired processing, but they are preceded.

### CODE REVIEW

**Comment by spencergibb** [comment]
Can you describe the problem you are facing that drove you to propose these changes?

**Comment by heowc** [comment]
> Can you describe the problem you are facing that drove you to propose these changes?

I am writing a pr that collects metrics using these events. However, the information reflected in the registry is unknown when the event occurs. This is because events are issued first before being registered/deleted.

```java
@EventListener
public void onEvent(EurekaInstanceRegisteredEvent event) {
    final InstanceInfo instanceInfo = registeredEvent.getInstanceInfo();
    instanceRegistry.getInstanceByAppAndId(instanceInfo.getAppName(), instanceInfo.getInstanceId());  <--
}
```

Also, I find it somewhat odd that events are issued before register/delete actions. Isn't that right?

**Comment by heowc** [comment]
> Can you describe the problem you are facing that drove you to propose these changes?

@spencergibb hello,

I am writing a pr that collects metrics using these events. However, the information reflected in the registry is unknown when the event occurs. This is because events are issued first before being registered/deleted.

```java
@EventListener
public void onEvent(EurekaInstanceRegisteredEvent event) {
    final InstanceInfo instanceInfo = event.getInstanceInfo();
    instanceRegistry.getInstanceByAppAndId(instanceInfo.getAppName(), instanceInfo.getInstanceId());  <--
}
```

Also, I find it somewhat odd that events are issued before register/delete actions. Isn't that right?

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java [hunk 1]
```diff
@@ -124,6 +121,17 @@ private void handleRegistration(InstanceInfo info, int leaseDuration, boolean is
 		publishEvent(new EurekaInstanceRegisteredEvent(this, info, leaseDuration, isReplication));
 	}
 
+	private void handleRenewal(final String appName, final String serverId, boolean isReplication) {
+		log("renew " + appName + " serverId " + serverId + ", isReplication {}" + isReplication);
```
Comment (OlgaMaciaszek) [inline spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java:125]: The logging should happen before renewing.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks for the PR, @heowc. Have added a comment. Please address.

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java [hunk 2]
```diff
@@ -78,43 +79,39 @@ public void openForTraffic(ApplicationInfoManager applicationInfoManager, int co
...
 	public void register(final InstanceInfo info, final boolean isReplication) {
-		handleRegistration(info, resolveInstanceLeaseDuration(info), isReplication);
 		super.register(info, isReplication);
+		handleRegistration(info, resolveInstanceLeaseDuration(info), isReplication);
 	}
 
 	@Override
 	public boolean cancel(String appName, String serverId, boolean isReplication) {
+		final boolean cancelled = super.cancel(appName, serverId, isReplication);
 		handleCancelation(appName, serverId, isReplication);
```
Comment (OlgaMaciaszek) [inline spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java:95]: Since we get a boolean, we might only want to  handle cancellation if `cancelled` is `true`.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java (hunk 1 above)
Comment (heowc) [reply spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java:125]: Is there a reason why renew is different?

Or should cancellation and registration be changed in that way? 🤔

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java (hunk 2 above)
Comment (heowc) [reply spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java:95]: That seems correct.
I'll change renew in the same way.

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java [hunk 3]
```diff
@@ -49,8 +50,8 @@ public class InstanceRegistry extends PeerAwareInstanceRegistryImpl implements A
 	private final int defaultOpenForTrafficCount;
 
 	public InstanceRegistry(EurekaServerConfig serverConfig, EurekaClientConfig clientConfig, ServerCodecs serverCodecs,
-			EurekaClient eurekaClient, EurekaServerHttpClientFactory eurekaServerHttpClientFactory,
-			int expectedNumberOfClientsSendingRenews, int defaultOpenForTrafficCount) {
+							EurekaClient eurekaClient, EurekaServerHttpClientFactory eurekaServerHttpClientFactory,
```
Comment (OlgaMaciaszek) [inline spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java:53]: Please remove unnecessary formatting change.

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java (hunk 1 above)
Comment (OlgaMaciaszek) [reply spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java:131]: No, sorry. I've misread; Please changed `"renew "` to "renewed "`, so that the sequence of events is clear.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks, @heowc. Looks good, but have added some cosmetic comments. Please take a look.

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java (hunk 1 above)
Comment (heowc) [reply spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java]: fixed :)

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java (hunk 3 above)
Comment (heowc) [reply spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java]: fixed :)

**Comment by heowc** [comment]
@OlgaMaciaszek,

Build failed. 😭 
But it works fine on my local.

Is this a flaky test?

https://github.com/spring-cloud/spring-cloud-netflix/actions/runs/8599692396/job/23571366157?pr=4274#step:4:25977

**Comment by OlgaMaciaszek** [comment]
It's not a flaky test - it's directly caused by the changes made. We are now checking for a boolean result of those operations that we're not getting (the test is not set up to run the full operation including communicating with an Eureka Server). You can stub it appropriately in your tests, but make sure to verify the behaviour against a running Eureka server in your local setup. It's bizarre that it passes for you locally - maybe some old code version cached somewhere?

**Comment by heowc** [comment]
> Please adjust tests to code changes.

Fixed 🤣

File: spring-cloud-netflix-eureka-server/src/test/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistryTests.java [hunk 4]
```diff
@@ -106,15 +106,7 @@ void testInternalCancel() {
 		// calling tested method
 		instanceRegistry.internalCancel(APP_NAME, HOST_NAME, false);
 		// event of proper type is registered
-		assertThat(this.testEvents.applicationEvents.size()).isEqualTo(1);
```
Comment (OlgaMaciaszek) [inline spring-cloud-netflix-eureka-server/src/test/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistryTests.java:109]: The idea is not to remove the assertions. Please keep the assersions, but add necessary stubbing so that `cancelled`/ `renewed` return `true`.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks @heowc, please see the comments and fix them.

File: spring-cloud-netflix-eureka-server/src/test/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistryTests.java (hunk 4 above)
Comment (heowc) [reply spring-cloud-netflix-eureka-server/src/test/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistryTests.java:109]: I agree. I changed this to registering the instance directly: It is too closely related to the internal implementation to utilize the stub object.

**Review by OlgaMaciaszek — APPROVED** [review]
LGTM.
